### PR TITLE
feat: remove usage of Timestamp in epoch inputs PE-7338

### PR DESCRIPTION
**Pull request by fedellen**

### CODE REVIEW

File: src/common/io.ts [hunk 1]
```diff
@@ -154,38 +154,52 @@ export class ARIOReadable implements AoARIORead {
...
+  private async computeEpochIndexForTimestamp(
+    timestamp: number,
+  ): Promise<number> {
+    const epochSettings = await this.getEpochSettings();
+    const epochZeroStartTimestamp = epochSettings.epochZeroStartTimestamp;
+    const epochLengthMs = epochSettings.durationMs;
+    return Math.floor((timestamp - epochZeroStartTimestamp) / epochLengthMs);
+  }
+
+  private async epochInputToTags(
```
Comment (fedellen) [inline src/common/io.ts:166]: nit: can return just one tag for no concat

Comment (fedellen) [reply src/common/io.ts:166]: gets complex cause we need to type check undefined, empty array is simpler. think its fine as is

File: src/common/io.ts [hunk 2]
```diff
@@ -154,38 +154,48 @@ export class ARIOReadable implements AoARIORead {
     });
   }
 
+  private async computeEpochIndexForTimestamp(
+    timestamp: number,
+  ): Promise<number> {
+    const epochSettings = await this.getEpochSettings();
```
Comment (dtfiedler) [inline src/common/io.ts:160]: you can memoize this by adding `private epochSettings: AoEpochSettings | undefined`. set it if it has not been set, return it if it already has.

File: src/common/io.ts [hunk 3]
```diff
@@ -409,16 +392,7 @@ export class ARIOReadable implements AoARIORead {
...
-        name: 'Timestamp',
-        value:
-          (epoch as { timestamp?: number })?.timestamp?.toString() ??
-          Date.now().toString(),
-      },
-      {
-        name: 'Epoch-Index',
-        value: (epoch as { epochIndex?: number })?.epochIndex?.toString(),
-      },
+      await this.epochInputToEpochIndexTag(epoch),
```
Comment (dtfiedler) [inline src/common/io.ts:395]: nit: pull this out of all tags to a var

File: src/common/io.ts [hunk 4]
```diff
@@ -154,38 +154,48 @@ export class ARIOReadable implements AoARIORead {
...
+    }
+    const timestamp = (params as { timestamp?: number })?.timestamp;
+    if (timestamp !== undefined) {
+      return {
+        name: 'Epoch-Index',
+        value: (await this.computeEpochIndexForTimestamp(timestamp)).toString(),
+      };
+    }
+    return { name: 'Epoch-Index', value: '' };
+  }
```
Comment (dtfiedler) [inline src/common/io.ts:182]:
```suggestion
  private async computeEpochIndex(
    params?: EpochInput,
  ): Promise<number | undefined> {
   const { epochIndex = undefined, timestamp = undefined } = params; // you can just ts-ignore
    if (timestamp){
        epochIndex ?= await await this.computeEpochIndexForTimestamp(timestamp)).toString()
    }
    return epochIndex
```
and then keep the `Epoch-Index` tag setup in the caller

File: src/common/io.ts (hunk 2 above)
Comment (dtfiedler) [reply src/common/io.ts:160]: correct, in that case it would be a forced refresh

File: src/common/io.ts [hunk 5]
```diff
@@ -154,39 +154,54 @@ export class ARIOReadable implements AoARIORead {
...
-        name: 'Timestamp',
-        value:
-          (params as { timestamp?: number })?.timestamp?.toString() ??
-          Date.now().toString(),
-      },
-      {
-        name: 'Epoch-Index',
-        value: (params as { epochIndex?: number })?.epochIndex?.toString(),
-      },
+      epochInputTag,
```
Comment (dtfiedler) [inline src/common/io.ts:192]: i don't believe this handler supports epoch-index, we don't store settings state on the epoch

Comment (fedellen) [reply src/common/io.ts]: [b311ca](https://github.com/ar-io/ar-io-sdk/pull/325/commits/1b311ca4d21c33ca5a6a44b942fb5e5673c09c37)

File: src/common/io.ts (hunk 4 above)
Comment (fedellen) [reply src/common/io.ts]: [42399a4](https://github.com/ar-io/ar-io-sdk/pull/325/commits/42399a42acc96b796edff30d03824034ea1c3e58)

File: src/common/io.ts (hunk 2 above)
Comment (fedellen) [reply src/common/io.ts:160]: memoized above getEpochSettings

File: src/common/io.ts [hunk 6]
```diff
@@ -154,38 +154,50 @@ export class ARIOReadable implements AoARIORead {
...
+
+    const timestamp = (params as { timestamp?: number })?.timestamp;
+    if (timestamp !== undefined) {
+      return (await this.computeEpochIndexForTimestamp(timestamp)).toString();
+    }
+
+    return undefined;
+  }
+
+  private epochSettings: AoEpochSettings | undefined;
```
Comment (dtfiedler) [inline src/common/io.ts:182]: nit, put this at the top of the class with other attributes - it gets lost down here

File: src/common/io.ts [hunk 7]
```diff
@@ -154,38 +154,50 @@ export class ARIOReadable implements AoARIORead {
...
+    }
+
+    const allTags = [{ name: 'Action', value: 'Epoch-Settings' }];
+
+    const epochSettings = await this.process.read<AoEpochSettings>({
       tags: pruneTags(allTags),
     });
+    this.epochSettings = epochSettings;
+    return epochSettings;
   }
```
Comment (dtfiedler) [inline src/common/io.ts:195]:
```suggestion
    this.epochSettings = await this.process.read<AoEpochSettings>({
      tags: pruneTags(allTags),
    });
    return this.epochSettings;
  }
```

**Comment by dtfiedler** [comment]
do we have e2e tests for these?

**Review by dtfiedler — APPROVED** [review]

**Review by dtfiedler — APPROVED** [review]

**Pull request merged**